PS-7778 Wrongly handled definer for triggers when PS used

i-rinat opened the pull request:
It's possible to get spurious access denied errors when a query causes a trigger to execute, and is executed as a prepared statement. Access checking logic tries to check if the current user has access to all tables that a trigger touches, even if that trigger is bound to run in a context of another user.

As all additional tables are added to the list after all own tables of a query, it's enough to stop traversing the table list once any of non-own tables are met.

This regression was introduced in 67c3c70e4895874d43434f1df556f9f30d781b48.

altmannmarcelo commented:
https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1035/

altmannmarcelo reviewed (changes requested):
Hi @i-rinat 

Can you please cover a few more tests cases:

1) when definer is different from user executing the trigger - trigger update a different table and definer lack privileges on update different table.
2) definer has no SELECT access on the table and trigger uses OLD.column NEW.column
3) definer has no UPDATE access on the table and trigger uses `SET NEW.column`
4) definer has TRIGGER privilege on the table, creates the trigger, then remove the TRIGGER privilege from definer user and execute the INSERT/UPDATE using a different user.
4) Repeat the above with PS and without PS

i-rinat commented:
@altmannmarcelo,

Sure, will do the changes.
Also, as far as I can understand, all additional test cases you mentioned should generate access errors. Is that right?

altmannmarcelo commented:
Correct. This is to validate we are not by-passing any requirement ACL restrictions with this patch.

kamil-holubicki commented on mysql-test/t/ps_trigger_access_check_regression.test:
```diff
@@ -0,0 +1,52 @@
+--echo #
+--echo # Fill tables.
+CREATE DATABASE test01;
+use test01;
```
Multiple occurrences in the file:
s/use/USE

i-rinat replied:
I've rewritten the test, and hopefully have all keywords in upper case now.

kamil-holubicki commented on mysql-test/t/ps_trigger_access_check_regression.test:
```diff
@@ -0,0 +1,52 @@
+
+--echo #
+--echo # Update should succeed.
+update tbl01 set a = now();
```
s/update/UPDATE
s/set/SET

i-rinat replied:
I've rewritten the test, and hopefully have all keywords in upper case now.

kamil-holubicki commented on mysql-test/t/ps_trigger_access_check_regression.test:
```diff
@@ -0,0 +1,52 @@
+update tbl01 set a = now();
+--echo #
+--echo # Same update as a prepare-execute pair should also succeed.
+prepare s from 'update tbl01 set a = now()';
```
From here to the end of the file: all SQL keywords should be upper case

i-rinat replied:
I've rewritten the test, and hopefully have all keywords in upper case now.

kamil-holubicki commented on sql/sql_select.cc:
```diff
@@ -946,9 +946,10 @@ bool Sql_cmd_select::check_privileges(THD *thd) {
 
-  for (TABLE_LIST *tr = lex->query_tables; tr != nullptr;
-       tr = tr->next_global) {
+  for (TABLE_LIST *tr = lex->query_tables;
```
Is only select affected, or there are others as well?

i-rinat replied:
This bug affect UPDATEs, INSERTs and DELETEs. It could hypothetically affect SELECTs, but as far as I know, it's not possible to create a trigger on SELECTs.

Although this code is in `sql_select.cc` file, it's the implementation of the `Sql_cmd_dml` class, so it's inherited by all DML commands. All descendant classes have own implementation of `<...>::check_privileges()`, but they all also call `Sql_cmd_dml::check_all_table_privileges()`, so changes here affect all DML commands.

i-rinat commented:
> cover a few more tests cases

Added mentioned test cases. Also rewritten the original test case to make it more readable. Additional test cases do generate errors as expected.

Tried to find if there are any other ways to trigger a similar bug. But as one can only attach a trigger only to insert, update, and delete operations, only those three may cause a similar issue. Handlers of all those commands call `Sql_cmd_dml::check_all_table_privileges()` during the privilege checking phase, so changes in `Sql_cmd_dml::check_all_table_privileges()` are enough to cover all cases.

altmannmarcelo reviewed (approved):
LGTM.
Also, I see the same logic (checking for first_not_own_table) been used and various other different places.
Jenkins looks good when compared to trunk job too.

altmannmarcelo commented:
@i-rinat can I ask you just to adjust the commit message:

```
Title: Fixed (JIRA KEY) - (JIRA Title)

JIRA LINK

Problem:
Description of the problem

Fix:
Description of the fix
```

One example from PXB - https://github.com/percona/percona-xtrabackup/pull/1172/commits/bf91d8989b1c12cb3c4e86b284a6de4dad57d1eb

i-rinat commented:
@altmannmarcelo

I've updated the commit message to include a description of the problem, why it happens, and how it's fixed. Could you please take a look?

Since the moment I initially created the pull request, `8.0` was updated. Do I need to rebase the patch onto the current state of `8.0`? Any other branch?

altmannmarcelo commented:
@i-rinat can you please rebase it on top of https://github.com/percona/percona-server/tree/release-8.0.26-16 and I will get it merged.

Thanks

i-rinat commented:
> rebase it on top of <...> release-8.0.26-16

Rebased.
